fcrepo/fcrepo
# Allow on demand reindexing of resources

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3553

# What does this Pull Request do?

Updates the ContainmentIndex and FedoraToOcflObjectIndex implementations to have `on conflict`/`on duplicate key` clauses in order for the reindexing to run on resources which already exist.

# How should this be tested?

* Create a resource in Fedora
* Update the resource outside of Fedora 
  * I'm not sure of the best way to do this -- I did it by hand which wasn't bad for a small resource but not really ideal with how many changes are needed for ocfl.
  * Ideally the updated resource should also have changes so that the `ReferenceService` is hit.
* Issue a POST to the reindex endpoint: `http://localhost:8080/rest/<fedora-id>/fcr:reindex`
* The reindex should succeed and any changes to the ContainmentIndex, FedoraToOCflObjectIndex, and ReferenceService should be visible.

# Notes

As I'm making this I remembered some discussion about limiting this based on the interaction model, which seems like something we would still want. I imagine there are ways to break this as it currently stands, so at least for an initial implementation we could add some restrictions before merging.

The changes to the `FedoraReindexIT` were fairly minimal, and the only check that's really done is that the http call returns a 204. It might be good to hit the endpoints to see changes in the resource being reflected.

# Interested parties

@fcrepo/committers
